boost 1.68 removed legacy support for sha1 in UUID

FreeBSD is quick to upgrade libraries and we are still using Boost's UUID SHA1 within `conversions.cpp` for calculating aggregated config (whole config) hashes.

Here are the release notes: https://www.boost.org/users/history/version_1_68_0.html

> Breaking change: sha1 detail namespace header redirection for backwards compatibility was removed (PR#69).

Here's an example of the build failure: https://jenkins.osquery.io/job/osqueryMasterBuildFreeBSD11/617/console

You may see something like:
```
[ 90%] Linking CXX static library libosquery_additional.a
[ 91%] Building CXX object osquery/CMakeFiles/libosquery.dir/config/parsers/views.cpp.o
[ 91%] Building CXX object osquery/CMakeFiles/libosquery.dir/core/base64.cpp.o
[ 91%] Building CXX object osquery/CMakeFiles/libosquery.dir/core/conversions.cpp.o
[ 91%] Building CXX object osquery/CMakeFiles/libosquery.dir/core/database/database.cpp.o
/var/jenkins/workspace/osqueryMasterBuildFreeBSD11/osquery/core/conversions.cpp:19:10: fatal error: 'boost/uuid/sha1.hpp' file not found
#include <boost/uuid/sha1.hpp>
         ^~~~~~~~~~~~~~~~~~~~~
[ 91%] Building CXX object osquery/CMakeFiles/libosquery.dir/core/database/in_memory_database.cpp.o
[ 92%] Building CXX object osquery/CMakeFiles/libosquery.dir/core/flags.cpp.o
[ 92%] Built target libosquery_additional
[ 92%] Building CXX object osquery/CMakeFiles/libosquery.dir/core/hashing.cpp.o
1 error generated.
gmake[4]: *** [osquery/CMakeFiles/libosquery.dir/build.make:232: osquery/CMakeFiles/libosquery.dir/core/conversions.cpp.o] Error 1
gmake[4]: *** Waiting for unfinished jobs....
gmake[3]: *** [CMakeFiles/Makefile2:1257: osquery/CMakeFiles/libosquery.dir/all] Error 2
gmake[2]: *** [CMakeFiles/Makefile2:1229: osquery/CMakeFiles/daemon.dir/rule] Error 2
gmake[1]: *** [Makefile:526: daemon] Error 2
gmake: *** [Makefile:323: daemon] Error 2
Build step 'Execute shell' marked build as failure
```

The fix is refactoring the SHA1 usage in Config to use the `hashing.cpp`/OpenSSL wrappers.

--- a/osquery/config/config.cpp
+++ b/osquery/config/config.cpp
@@ -30,6 +30,7 @@
 
 #include "osquery/core/conversions.h"
 #include "osquery/core/flagalias.h"
+#include "osquery/core/hashing.h"
 
 namespace rj = rapidjson;
 
@@ -898,7 +899,9 @@ void Config::getPerformanceStats(
 }
 
 bool Config::hashSource(const std::string& source, const std::string& content) {
-  auto new_hash = getBufferSHA1(content.c_str(), content.size());
+  Hash hash(HASH_TYPE_SHA1);
+  hash.update(content.c_str(), content.size());
+  auto new_hash = hash.digest();
 
   WriteLock wlock(config_hash_mutex_);
   if (hash_[source] == new_hash) {
@@ -924,7 +927,10 @@ Status Config::genHash(std::string& hash) const {
   for (const auto& it : hash_) {
     add(it.second);
   }
-  hash = getBufferSHA1(buffer.data(), buffer.size());
+
+  Hash new_hash(HASH_TYPE_SHA1);
+  new_hash.update(buffer.data(), buffer.size());
+  hash = new_hash.digest();
 
   return Status(0, "OK");
 }

--- a/osquery/config/packs.cpp
+++ b/osquery/config/packs.cpp
@@ -19,6 +19,7 @@
 #include <osquery/system.h>
 
 #include "osquery/core/conversions.h"
+#include "osquery/core/hashing.h"
 #include "osquery/core/json.h"
 
 namespace rj = rapidjson;
@@ -71,7 +72,10 @@ size_t getMachineShard(const std::string& hostname = "", bool force = false) {
 
   // An optional input hostname may override hostname detection for testing.
   auto hn = (hostname.empty()) ? getHostname() : hostname;
-  auto hn_hash = getBufferSHA1(hn.c_str(), hn.size());
+
+  Hash hash(HASH_TYPE_SHA1);
+  hash.update(hn.c_str(), hn.size());
+  auto hn_hash = hash.digest();
 
   if (hn_hash.size() >= 2) {
     auto const hn_num = tryTo<long>(hn_hash.substr(0, 2), 16);

--- a/osquery/config/tests/config_tests.cpp
+++ b/osquery/config/tests/config_tests.cpp
@@ -240,23 +240,28 @@ TEST_F(ConfigTests, test_pack_removal) {
 }
 
 TEST_F(ConfigTests, test_content_update) {
+  const std::string source{"awesome"};
+
   // Read config content manually.
   std::string content;
   readFile(kTestDataPath + "test_parse_items.conf", content);
 
   // Create the output of a `genConfig`.
   std::map<std::string, std::string> config_data;
-  config_data["awesome"] = content;
+  config_data[source] = content;
 
   // Update, then clear, packs should have been cleared.
   get().update(config_data);
+  auto source_hash = get().getHash(source);
+  EXPECT_EQ("29d117ea900322c88e85e349db01ee386727a484", source_hash);
+
   size_t count = 0;
   auto packCounter = [&count](const Pack& pack) { count++; };
   get().packs(packCounter);
   EXPECT_GT(count, 0U);
 
   // Now clear.
-  config_data["awesome"] = "";
+  config_data[source] = "";
   get().update(config_data);
   count = 0;
   get().packs(packCounter);

--- a/osquery/config/tests/config_tests.cpp
+++ b/osquery/config/tests/config_tests.cpp
@@ -253,7 +253,12 @@ TEST_F(ConfigTests, test_content_update) {
   // Update, then clear, packs should have been cleared.
   get().update(config_data);
   auto source_hash = get().getHash(source);
-  EXPECT_EQ("29d117ea900322c88e85e349db01ee386727a484", source_hash);
+  // TODO(#5069/#5070) unsure why this value is different on Windows.
+  if (isPlatform(PlatformType::TYPE_WINDOWS)) {
+    EXPECT_EQ("c755b2762de2ef45415972cc3c2af15109f16f1c", source_hash);
+  } else {
+    EXPECT_EQ("29d117ea900322c88e85e349db01ee386727a484", source_hash);
+  }
 
   size_t count = 0;
   auto packCounter = [&count](const Pack& pack) { count++; };

--- a/osquery/core/conversions.cpp
+++ b/osquery/core/conversions.cpp
@@ -12,6 +12,8 @@
 #include <locale>
 #include <unordered_map>
 
+#include <boost/io/detail/quoted_manip.hpp>
+
 #include <osquery/logger.h>
 
 #include "osquery/core/conversions.h"

--- a/osquery/core/conversions.cpp
+++ b/osquery/core/conversions.cpp
@@ -12,13 +12,6 @@
 #include <locale>
 #include <unordered_map>
 
-#include <boost/io/detail/quoted_manip.hpp>
-#if (BOOST_VERSION >= 106600)
-#include <boost/uuid/detail/sha1.hpp>
-#else
-#include <boost/uuid/sha1.hpp>
-#endif
-
 #include <osquery/logger.h>
 
 #include "osquery/core/conversions.h"
@@ -351,22 +344,6 @@ std::vector<std::string> split(const std::string& s,
     elems.push_back(boost::algorithm::join(accumulator, delims));
   }
   return elems;
-}
-
-std::string getBufferSHA1(const char* buffer, size_t size) {
-  // SHA1 produces 160-bit digests, so allocate (5 * 32) bits.
-  uint32_t digest[5] = {0};
-  boost::uuids::detail::sha1 sha1;
-  sha1.process_bytes(buffer, size);
-  sha1.get_digest(digest);
-
-  // Convert digest to desired hex string representation.
-  std::stringstream result;
-  result << std::hex << std::setfill('0');
-  for (size_t i = 0; i < 5; ++i) {
-    result << std::setw(sizeof(uint32_t) * 2) << digest[i];
-  }
-  return result.str();
 }
 
 size_t operator"" _sz(unsigned long long int x) {

--- a/osquery/core/conversions.h
+++ b/osquery/core/conversions.h
@@ -138,15 +138,6 @@ inline size_t utf8StringSize(const std::string& str) {
   return res;
 }
 
-/*
- * @brief Request a SHA1 hash from the contents of a buffer.
- *
- * @param buffer A caller-controlled buffer (already allocated).
- * @param size The length of the controlled buffer.
- * @return A string (hex) representation of the hash digest.
- */
-std::string getBufferSHA1(const char* buffer, size_t size);
-
 #ifdef DARWIN
 /**
  * @brief Convert a CFStringRef to a std::string.

--- a/osquery/core/tests/conversions_tests.cpp
+++ b/osquery/core/tests/conversions_tests.cpp
@@ -75,12 +75,6 @@ TEST_F(ConversionsTests, test_split_occurences) {
   EXPECT_EQ(split(content, ':', 1), expected);
 }
 
-TEST_F(ConversionsTests, test_buffer_sha1) {
-  std::string test = "test\n";
-  EXPECT_EQ("4e1243bd22c66e76c2ba9eddc1f91394e57f9f83",
-            getBufferSHA1(test.c_str(), test.size()));
-}
-
 TEST_F(ConversionsTests, test_json_array) {
   auto doc = JSON::newArray();
 

